### DISCUSSION TITLE
Gas transmission network modelling: How can I add compressors and flow resistance?

### DISCUSSION THREAD

**Discussion by ananthu2050** [e2nIEE/pandapipes#411] [Q&A]
I am interested in using pandapipes for gas transportation networks. In such networks, we have compressor stations, injection points, flow controllers etc in addition to pipes, junctions, valves that are currently available for modelling in pandapipes. To model the gas transmission networks, I need these additional equipment. I am not sure how to add new component like compressor in pandapipes. I looked at standard type library notebook and currently I could see pipes and pumps. I need  some guidance for adding compressors, flow regulators etc in standard type library. Appreciate any inputs for the same.

**Comment by jkisse** [comment]
Hi ananthu2050,
the components include the physical model, the necessary equations for the Newton Raphson etc.. For all components, there are create...-functions that have the relevant topological parameters (mainly from/to junctions) and model parameters (e.g., diameter, roughness, open/close status, ...) as inputs. 
On top of that, the standard types are simply a way to use these create...-functions more conveniently by offering pre-defined sets of parameters, e.g. for nominal pipe types.

Recently, the new components "[compressor](https://github.com/e2nIEE/pandapipes/blob/develop/pandapipes/component_models/compressor_component.py)" and "[pressure controller](https://github.com/e2nIEE/pandapipes/blob/develop/pandapipes/component_models/pressure_control_component.py)" have been added to pandapipes.
The compressor is based on a simple model that increases the outlet pressure by a fixed ratio compared to the inlet pressure. (This ratio could be dynamically modified during runtime by adding "Controllers"). 
The pressure controller introduces a pressure drop between the from- and to_junction so that a pre-defined pressure at a "control junction" downstream is met.
There is also the "external grid" component which fixes the pressure at the junction it is connected to and has a variable (slack) mass flow.
@dlohmeier is also developing a regulator that creates a pressure difference to meet a pre-defined volume flow (flow regulator) on his branch "[flow_control](https://github.com/dlohmeier/pandapipes/blob/flow_control/pandapipes/component_models/flow_control_component.py)" that will be merged into the main pandapipes.

I hope this helps you to get started. Could you please specify further what you have in mind (which level of detail) for the compressor stations and injection points?

**Comment by ananthu2050** [comment]
Thank you for the super quick response. "[compressor](https://github.com/e2nIEE/pandapipes/blob/develop/pandapipes/component_models/compressor_component.py)" and "[pressure controller](https://github.com/e2nIEE/pandapipes/blob/develop/pandapipes/component_models/pressure_control_component.py)" would be good starting points for me. The compressor is driven by a motor fueled by some of the natural gas bled from the pipeline. So there is also small difference in the volume of the gas before and after the compressor station. I think, I will have to implement the same in the compressor component. Also compressor stations will have some scrubbers or strainers to remove impurities or liquids. They might also cause some pressure drop, but I believe that would be negligible. That could be the other level of detail that I need to implement. But I think pressure controller can be readily used.  "[flow_control](https://github.com/dlohmeier/pandapipes/blob/flow_control/pandapipes/component_models/flow_control_component.py)" would be useful as well.

**Comment by jkisse** [comment]
The consumption of the compressor can be modelled by adding a sink at either inlet or outlet junction and controlling it via a ["Controller" ](https://pandapower.readthedocs.io/en/latest/control.html) from pandapower. The Controller object is executed in every iteration step of the pipeflow calculation and will manipulate one variable in the net (mass flow of the compressor sink) based on another variable (compression power). The pandapower Controller is so general that it should be possible to use it in pandapipes without modifications.
Regarding the pressure drop in the compressor station, you could use an open [valve ](https://pandapipes.readthedocs.io/en/develop/components/valve/valve_component.html) with loss coefficient > 0 or merge that loss coefficient behavior into the compressor component, i.e. adding the respective to the compressors pit-entries. We do not have a dedicated "resistor" component yet, but it should not be too hard to develop something based on the existing (abstract) components.

If you develop new components or extend the existing ones, feel free to open a pull request : )

**Comment by ananthu2050** [comment]
Thank you @jkisse I will open a pull request to implement the changes.
